cmaes/_safe_cma.py

```diff
@@ -0,0 +1,702 @@
+
+            # safety function
+            def safe_function(x):
+                return x[0]
+
+            # safe seeds
+            safe_seeds_num = 10
+            safe_seeds = (np.random.rand(safe_seeds_num, dim) * 2 - 1) * 5
+            safe_seeds[:, 0] = - np.abs(safe_seeds[:, 0])
+
+            # evaluation of safe seeds (with multiple safety functions)
+            seeds_evals = np.array([quadratic(x) for x in safe_seeds])
+            seeds_safe_evals = np.stack([[safe_function(x)] for x in safe_seeds])
+            safety_threshold = np.array([0])
+
+            # optimizer (safe CMA-ES)
+            optimizer = SafeCMA(
+                sigma=1.,
+                safety_threshold=safety_threshold,
+                safe_seeds=safe_seeds,
+                seeds_evals=seeds_evals,
+                seeds_safe_evals=seeds_safe_evals,
+            )
+
+            unsafe_eval_counts = 0
+            best_eval = np.inf
+
+            for generation in range(400):
+                solutions = []
+                for _ in range(optimizer.population_size):
+                    # Ask a parameter
+                    x = optimizer.ask()
+                    value = quadratic(x)
+                    safe_value = np.array([safe_function(x)])
+
+                    # save best eval
+                    best_eval = np.min((best_eval, value))
+                    unsafe_eval_counts += (safe_value > safety_threshold)
+
+                    solutions.append((x, value, safe_value))
+
+                # Tell evaluation values.
+                optimizer.tell(solutions)
+
+                print(f"#{generation} ({best_eval} {unsafe_eval_counts})")
+
+                if optimizer.should_stop():
+                    break
+
+    Args:
+
+        safe_seeds:
+            Solutions whose safe function values are above the safety thresholds.
+            Safe CMA-ES uses the safe seed with the best evaluation value as
+            the initial mean vector of multi-variate Gaussian distributions.
+
+        seeds_evals:
+            Evaluation values of safe seeds on the objective function.
+
+        seeds_safe_evals:
+            Evaluation values of safe seeds on the safe functions.
+
+        safety_threshold:
+            Safety thresholds for each safe functions.
+
+        sigma:
+            Initial standard deviation of covariance matrix.
+            Safe CMA-ES modifies sigma when more than two safe seeds are given.
+
+        bounds:
+            Lower and upper domain boundaries for each parameter (optional).
+
+        n_max_resampling:
+            A maximum number of resampling parameters (default: 100).
+            If all sampled parameters are infeasible, the last sampled one
+            will be clipped with lower and upper bounds.
+
+        seed:
+            A seed number (optional).
+
+        population_size:
+            A population size (optional).
+
+        cov:
+            A covariance matrix (optional).
+    """
```

Comment:
Please add the paper link like [this](https://github.com/CyberAgentAILab/cmaes/blob/main/cmaes/_catcma.py#L81).